crypto/modes/asm/aes-gcm-avx512.pl

```diff
@@ -0,0 +1,5141 @@
+  "ghash");
+&GCM_UPDATE_AAD(
+  "$arg1",  "$arg2",  "$arg3",  "%r10",   "%r11",   "%r12",  "%k1",   "%xmm14", "%zmm1",  "%zmm11",
+  "%zmm3",  "%zmm4",  "%zmm5",  "%zmm6",  "%zmm7",  "%zmm8", "%zmm9", "%zmm10", "%zmm12", "%zmm13",
+  "%zmm15", "%zmm16", "%zmm17", "%zmm18", "%zmm19", "%zmm20");
+&EPILOG(
+  1,    # hkeys were allocated
+  $arg3);
+$code .= <<___;
+.Lexit_update_aad:
+ret
+.Lghash_seh_end:
+.cfi_endproc
+.size ossl_aes_gcm_update_aad_avx512, .-ossl_aes_gcm_update_aad_avx512
+___
+
+# ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+# ;void   ossl_aes_gcm_encrypt_avx512
+# ;       (const void* aes_keys,
+# ;        void *gcm128ctx,
+# ;        unsigned int *pblocklen,
+# ;        const unsigned char *in,
+# ;        size_t len,
+# ;        unsigned char *out);
+# ;
+# ; Performs encryption of data |in| of len |len|, and stores the output in |out|.
+# ; Stores encrypted partial block (if any) in gcm128ctx and its length in |pblocklen|.
+# ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+$code .= <<___;
+.globl ossl_aes_gcm_encrypt_avx512
+.type ossl_aes_gcm_encrypt_avx512,\@abi-omnipotent
+.align 32
+ossl_aes_gcm_encrypt_avx512:
+.cfi_startproc
+.Lencrypt_seh_begin:
+        endbranch
+___
+
+# ; NOTE: code before PROLOG() must not modify any registers
+&PROLOG(
+  1,    # allocate stack space for hkeys
+  1,    # allocate stack space for AES blocks
+  "encrypt");
+if ($CHECK_FUNCTION_ARGUMENTS) {
+  $code .= <<___;
+        # ;; Check aes_keys != NULL
+        test               $arg1,$arg1
+jz      .Lexit_gcm_encrypt
+
+        # ;; Check gcm128ctx != NULL
+        test               $arg2,$arg2
+jz      .Lexit_gcm_encrypt
+
+        # ;; Check pblocklen != NULL
+        test               $arg3,$arg3
+jz      .Lexit_gcm_encrypt
+
+        # ;; Check in != NULL
+        test               $arg4,$arg4
+jz      .Lexit_gcm_encrypt
+
+        # ;; Check if len != 0
+        cmp                \$0,$arg5
+jz      .Lexit_gcm_encrypt
+
+        # ;; Check out != NULL
+        cmp                \$0,$arg6
+jz      .Lexit_gcm_encrypt
+___
+}
+$code .= <<___;
+        # ; load number of rounds from AES_KEY structure (offset in bytes is
+        # ; size of the |rd_key| buffer)
+        mov             `4*15*4`($arg1),%eax
+        cmp             \$9,%eax
+        je              .Laes_gcm_encrypt_128_avx512
+        cmp             \$11,%eax
+        je              .Laes_gcm_encrypt_192_avx512
+        cmp             \$13,%eax
+        je              .Laes_gcm_encrypt_256_avx512
+        xor             %eax,%eax
+        jmp             .Lexit_gcm_encrypt
+___
+for my $keylen (sort keys %aes_rounds) {
+  $NROUNDS = $aes_rounds{$keylen};
+  $code .= <<___;
+.align 32
+.Laes_gcm_encrypt_${keylen}_avx512:
+___
+  &GCM_ENC_DEC("$arg1", "$arg2", "$arg3", "$arg4", "$arg5", "$arg6", "ENC", "multi_call");
```

Comment:
Look like all macros are called with "multi-call". Is there a reason to keep all the "single-call" code in the macros?